### PR TITLE
Stacks 2.05: Allow epochs to be specified in Toml for Testnet

### PR DESCRIPTION
**Pull request by gregorycoppola**
## Description
We add deserialization to `BurnchainConfig` so this can be specified in the Toml file.

We allow a custom value for `epochs` to be used, unless the network is Mainnet.

**This allows us to specify custom epoch schedule for the testnet.**

## Testing

I tested this on the private test net (`github.com:blockstack/stacks-local-dev.git`).

Using the following epochs, I observed it move from `Epoch20` to `Epoch2_05`:
```
[[burnchain.epochs]]
epoch_id = 'Epoch20'
start_height = 0
end_height = 370

[burnchain.epochs.block_limit]
write_length = 100000000
write_count = 1000
read_length = 1000000000
read_count = 15000
runtime = 5000000000

[[burnchain.epochs]]
epoch_id = 'Epoch2_05'
start_height = 370
end_height = 9223372036854775807

[burnchain.epochs.block_limit]
write_length = 100000000
write_count = 2000
read_length = 1000000000
read_count = 50000
runtime = 5000000000
```

### CODE REVIEW

**Comment by gregorycoppola** [comment]
I'm still testing this but can be reviewed for the approach.

File: src/chainstate/burn/db/sortdb.rs [hunk 1]
```diff
@@ -2240,7 +2240,7 @@ impl SortitionDB {
         first_burn_header_timestamp: u64,
         epochs_ref: &[StacksEpoch],
     ) -> Result<(), db_error> {
-        debug!("Instantiate SortDB");
+        info!("Instantiating SortDB with epochs {:?}", &epochs_ref);
```
Comment (kantai) [inline src/chainstate/burn/db/sortdb.rs:2243]: This should remain a debug.

**Review by pavitthrap — APPROVED** [review]
lgtm - thanks for adding comments!

File: testnet/stacks-node/src/config.rs [hunk 2]
```diff
@@ -919,6 +923,8 @@ pub struct BurnchainConfig {
     pub leader_key_tx_estimated_size: u64,
     pub block_commit_tx_estimated_size: u64,
     pub rbf_fee_increment: u64,
+    // Custom override for the definitions of the epochs. This will only be applies for testnet and
+    // regtest nodes.
```
Comment (kantai) [inline testnet/stacks-node/src/config.rs:927]: `applies` -> `applied`

Also, might as well make that a rustdoc.

File: src/burnchains/bitcoin/indexer.rs [hunk 3]
```diff
@@ -706,9 +706,18 @@ impl BurnchainIndexer for BitcoinIndexer {
...
+    /// 2) Use hard-coded static values, otherwise.
+    ///
+    /// It is an error (panic) to set custom epochs if running on `Mainnet`.
     fn get_stacks_epochs(&self) -> Vec<StacksEpoch> {
         match self.config.epochs {
-            Some(ref epochs) => epochs.clone(),
+            Some(ref epochs) => {
+                assert!(self.runtime.network_id != BitcoinNetworkType::Mainnet);
+                epochs.clone()
+            }
```
Comment (kantai) [inline src/burnchains/bitcoin/indexer.rs:720]: This seems like the wrong place to panic -- the panic should occur earlier, when the configuration is first applied (as in somewhere in `testnet/stacks-node/src/config` or the `neon` run loop). I think leave this assertion in place, but just add another one that occurs earlier.

**Review by kantai — APPROVED** [review]
This LGTM -- just a couple small comments.

File: src/burnchains/bitcoin/indexer.rs (hunk 3 above)
Comment (jcnelson) [reply src/burnchains/bitcoin/indexer.rs:720]: Seconded.  Never put off until runtime that which can be handled at boot time.

File: testnet/stacks-node/src/config.rs [hunk 4]
```diff
@@ -895,7 +899,7 @@ impl std::default::Default for Config {
     }
 }
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
```
Comment (jcnelson) [inline testnet/stacks-node/src/config.rs:902]: Why does this need `Serialize` and `Deserialize`?

File: testnet/stacks-node/src/config.rs [hunk 5]
```diff
@@ -979,7 +985,7 @@ impl BurnchainConfig {
     }
 }
 
-#[derive(Clone, Deserialize, Default)]
+#[derive(Clone, Serialize, Deserialize, Default, Debug)]
```
Comment (jcnelson) [inline testnet/stacks-node/src/config.rs:988]: Why does this need `Serialize`?

File: testnet/stacks-node/src/config.rs [hunk 6]
```diff
@@ -1348,7 +1355,7 @@ impl MinerConfig {
     }
 }
 
-#[derive(Clone, Default, Deserialize)]
+#[derive(Clone, Default, Deserialize, Debug)]
```
Comment (jcnelson) [inline testnet/stacks-node/src/config.rs:1358]: Why do all of these structs here need `Debug`?

**Review by jcnelson — CHANGES_REQUESTED** [review]
Please address the panic feedback before merging, and please remove unused `#[derive(...)]`s.

File: testnet/stacks-node/src/config.rs (hunk 6 above)
Comment (gregorycoppola) [reply testnet/stacks-node/src/config.rs]: I did that so we can print the config files to screen. I put a part in the PR description about this.

However I'll revert.

File: testnet/stacks-node/src/config.rs (hunk 5 above)
Comment (gregorycoppola) [reply testnet/stacks-node/src/config.rs]: I removed the Serialize's

File: testnet/stacks-node/src/config.rs (hunk 4 above)
Comment (gregorycoppola) [reply testnet/stacks-node/src/config.rs]: removed serialize

File: src/chainstate/burn/db/sortdb.rs (hunk 1 above)
Comment (gregorycoppola) [reply src/chainstate/burn/db/sortdb.rs]: reverted

File: src/burnchains/bitcoin/indexer.rs (hunk 3 above)
Comment (gregorycoppola) [reply src/burnchains/bitcoin/indexer.rs:720]: I put a check in `BitcoinRegtestController::with_burnchain`.

**Review by gregorycoppola — COMMENTED** [review]
thanks.

* fail earlier on misconfiguration
* removed the Serialize's
* removed the Debug's

**Review by jcnelson — APPROVED** [review]
LGTM!

**Pull request merged**